### PR TITLE
feat: Show try operator propogated types on ranged hover 

### LINKED ISSUE #9931: Hover over `?` should show the type conversion
image

Would be cool to show `BarError from FooError` here! 

cc @Veykril

### PR DESCRIPTION
**Pull request by Veykril**
Basically this just shows the type of the inner expression of the `?` expression as well as the type of the expression that the `?` returns from:
Code_wIrCxMqLH9

Unless both of these types are `core::result::Result` in which case we show the error types only.
Code_Xruw5FCBNI

If both types are `core::option::Option` with different type params we do not show this special hover either as it would be pointless(instead fallback to default type hover)

Very much open to changes to the hover text here(I suppose we also want to show the actual type of the `?` expression, that is its output type?).

Fixes #9931

### CODE REVIEW

File: crates/ide/src/hover.rs [hunk 1]
```diff
@@ -242,15 +241,109 @@ fn hover_ranged(
...
+    let body_ty = body_ty.display(sema.db).to_string();
+    let ty_len_max = inner_ty.len().max(body_ty.len());
+
+    // "Propagated as: ".len() - " Type: ".len() = 8
+    let static_test_len_diff = 8 - s.len() as isize;
+    let tpad = static_test_len_diff.max(0) as usize;
+    let ppad = static_test_len_diff.min(0).abs() as usize;
+
+    res.markup = format!(
+        "{backticks}text\n{} Type: {:>pad0$}\nPropagated as: {:>pad1$}\n{backticks}\n",
```
Comment (Veykril) [inline crates/ide/src/hover.rs:335]: Note to self to fix later, the `text` string is lonely here with markdown disabled

File: crates/hir_ty/src/infer/pat.rs [hunk 2]
```diff
@@ -10,9 +9,10 @@ use hir_def::{
 };
 use hir_expand::name::Name;
 
-use super::{BindingMode, Expectation, InferenceContext, TypeMismatch};
 use crate::{
-    infer::{Adjust, Adjustment, AutoBorrow},
+    infer::{
+        Adjust, Adjustment, AutoBorrow, BindingMode, Expectation, InferenceContext, TypeMismatch,
+    },
```
Comment (matklad) [inline crates/hir_ty/src/infer/pat.rs:15]: :+1: drive by changes are ok if they make code consistent with the style guide!

File: crates/ide/src/hover.rs [hunk 3]
```diff
@@ -242,15 +241,110 @@ fn hover_ranged(
...
+
+    let mut inner_ty = inner_ty;
+    let mut s = "Try Target".to_owned();
+
+    let adts = inner_ty.as_adt().zip(body_ty.as_adt());
+    if let Some((hir::Adt::Enum(inner), hir::Adt::Enum(body))) = adts {
+        let famous_defs = FamousDefs(sema, sema.scope(&try_expr.syntax()).krate());
+        // special case for two options, there is no value in showing them
+        if let Some(option_enum) = famous_defs.core_option_Option() {
+            if inner == option_enum && body == option_enum {
```
Comment (matklad) [inline crates/ide/src/hover.rs:294]: cov_mark?

File: crates/ide/src/hover.rs [hunk 4]
```diff
@@ -242,15 +241,110 @@ fn hover_ranged(
...
+    };
+    walk_and_push_ty(sema.db, &inner_ty, &mut push_new_def);
+    walk_and_push_ty(sema.db, &body_ty, &mut push_new_def);
+    res.actions.push(HoverAction::goto_type_from_targets(sema.db, targets));
+
+    let inner_ty = inner_ty.display(sema.db).to_string();
+    let body_ty = body_ty.display(sema.db).to_string();
+    let ty_len_max = inner_ty.len().max(body_ty.len());
+
+    // "Propagated as: ".len() - " Type: ".len() = 8
```
Comment (matklad) [inline crates/ide/src/hover.rs:329]:
```suggestion
    let l = "Propagated as: ".len() - " Type: ".len();
```

**Review by matklad — COMMENTED** [review]
Looks plausible to me!

One potentail issue here is that `hover` is the largest (in terms of binary size) function in the ide crate. We might want to try to add mode software engineering here, and split this up in modules. 

It does seem that hover works roughly by "let's compute this n bits of info, and then combine then into one string". It would be cool if each of n bit was serpatate module, and if they returned some rust Structs with hir types, with the top-level doing actual rendering.

r=me

**Comment by matklad** [comment]
While we are at it...

image

This should mention deref somehow. BTW, do we have an official name for winking pretzel `&*` ?

**Comment by Veykril** [comment]
Another thing, we should also probably mention iterator item types for iterator types as well.

> BTW, do we have an official name for winking pretzel `&*`?
I only know it as a reborrow, haven't heard a specific name like that for it yet 😄

**Comment by Veykril** [comment]
bors r=matklad

**Pull request merged**